include/oneapi/dpl/internal/scan_by_segment_impl.h

```diff
@@ -0,0 +1,372 @@
+            ::std::forward<_ExecutionPolicy>(__exec));
+        auto __kernel2 = __par_backend_hetero::__internal::__kernel_compiler<_SegScanKernel2>::__compile(
+            ::std::forward<_ExecutionPolicy>(__exec));
+        __wgroup_size = ::std::min({ __wgroup_size, oneapi::dpl::__internal::__kernel_work_group_size(
+            ::std::forward<_ExecutionPolicy>(__exec), __kernel1), oneapi::dpl::__internal::__kernel_work_group_size(
+            ::std::forward<_ExecutionPolicy>(__exec), __kernel2) });
+#endif
+
+        ::std::size_t __n_groups = __par_backend_hetero::__ceiling_div(__n, __wgroup_size * __vals_per_item);
+
+        auto __partials =
+            oneapi::dpl::__par_backend_hetero::__internal::__buffer<_ExecutionPolicy, __val_type>(__exec, __n_groups)
+                .get_buffer();
+
+        // the number of segment ends found in each work group
+        auto __seg_ends =
+            oneapi::dpl::__par_backend_hetero::__internal::__buffer<_ExecutionPolicy, __flag_type>(__exec, __n_groups)
+                .get_buffer();
+
+        // 1. Work group reduction
+        auto __wg_scan = __exec.queue().submit([&](sycl::handler& __cgh) {
+            auto __partials_acc = __partials.template get_access<sycl::access_mode::write>(__cgh);
+            auto __seg_ends_acc = __seg_ends.template get_access<sycl::access_mode::write>(__cgh);
+
+            oneapi::dpl::__ranges::__require_access(__cgh, __keys, __values, __out_values);
+
+            __dpl_sycl::__local_accessor<__val_type> __loc_acc(2 * __wgroup_size, __cgh);
+
+#if _ONEDPL_COMPILE_KERNEL && _ONEDPL_KERNEL_BUNDLE_PRESENT
+            __cgh.use_kernel_bundle(__kernel1.get_kernel_bundle());
+#endif
+            __cgh.parallel_for<_SegScanKernel1>(
+#if _ONEDPL_COMPILE_KERNEL && !_ONEDPL_KERNEL_BUNDLE_PRESENT
+                __kernel1,
+#endif
+                sycl::nd_range<1>{__n_groups * __wgroup_size, __wgroup_size}, [=](sycl::nd_item<1> __item) {
+                    __val_type __accumulator = __identity;
+
+                    auto __group = __item.get_group();
+                    ::std::size_t __global_id = __item.get_global_id();
+                    ::std::size_t __local_id = __group.get_local_id();
+
+                    // 1a. Perform a serial scan within the work item over assigned elements. Store partial
+                    // reductions in work item memory, and write the accumulated value and number of counted
+                    // segments into work group memory.
+                    ::std::size_t __start = __global_id * __vals_per_item;
+                    ::std::size_t __end = __dpl_sycl::__minimum<decltype(__n)>{}(__start + __vals_per_item, __n);
+
+                    // First work item must set their accumulator to the provided init
+                    if (__global_id == 0 && __local_id == 0)
+                    {
+                        __accumulator = __init;
+                    }
+
+                    const int64_t NO_SEGMENT_BREAK = -1;
```

Comment:
It probably doesn't matter, but this `const` can be `constexpr`. Also `int64_t` -> `::std::int64_t`.

Reply:
I will make these changes.